core/src/state/world_state.rs

```diff
@@ -65,44 +66,94 @@ impl StateHandle {
+will return immediately on all its dependent notifies.
+*/
+#[derive(Debug)]
+pub struct ClosableNotify {
+    notifier: Arc<Sender<()>>,
+    rx: Receiver<()>,
+}
 
-    /** Block asynchronously until the given sequence number is reached. */
-    pub async fn wait_for_seq(&self, sequence: u64) {
-        if self.state().logical_time >= sequence {
-            return;
+impl ClosableNotify {
+    fn new() -> Self {
+        let (tx, rx) = tokio::sync::broadcast::channel(128);
+        Self {
+            notifier: Arc::new(tx),
+            rx,
         }
-        let receiver = { self.write_state().get_listener(sequence) };
-        receiver.notified().await;
+    }
+
+    pub async fn notified(&mut self) {
+        let _ = self.rx.recv().await;
+    }
+
+    fn notify_waiters(&self) {
+        let _ = self.notifier.send(());
     }
 }
 
+impl Clone for ClosableNotify {
+    fn clone(&self) -> Self {
+        Self {
+            notifier: self.notifier.clone(),
+            rx: self.notifier.subscribe(),
+        }
+    }
+}
 #[derive(Default, Debug)]
 pub struct WorldState {
     logical_time: u64,
     pub clusters: BTreeMap<ClusterName, ClusterState>,
-    listeners: HashMap<u64, Arc<Notify>>,
+    listeners: RwLock<BTreeMap<u64, ClosableNotify>>,
 }
 
 impl WorldState {
-    pub fn get_listener(&mut self, sequence: u64) -> Arc<Notify> {
-        match self.listeners.entry(sequence) {
-            Entry::Occupied(entry) => entry.get().clone(),
-            Entry::Vacant(entry) => {
-                let notify = Arc::new(Notify::new());
+    pub fn logical_time(&self) -> u64 {
+        self.logical_time
+    }
+
+    pub fn get_listener(&self, sequence: u64) -> ClosableNotify {
+        if self.logical_time >= sequence {
+            tracing::info!("immediately returning notify created");
+            let notify = ClosableNotify::new();
+            // this makes it so the notify returns immediately.
+            notify.notify_waiters();
+            return notify;
+        }
+        match self.listeners.write().unwrap().entry(sequence) {
+            std::collections::btree_map::Entry::Occupied(entry) => entry.get().clone(),
+            std::collections::btree_map::Entry::Vacant(entry) => {
+                let notify = ClosableNotify::new();
                 entry.insert(notify.clone());
                 notify
             }
         }
     }
 
     pub fn apply(&mut self, message: WorldStateMessage, sequence: u64) {
+        tracing::info!(?message, ?sequence, "World state message!");
         let cluster = self.clusters.entry(message.cluster.clone()).or_default();
         cluster.apply(message.message);
         self.logical_time = sequence;
 
-        if let Some(sender) = self.listeners.remove(&sequence) {
-            sender.notify_waiters();
-        }
+        let mut listeners = self.listeners.write().unwrap();
+        *listeners = {
+            let b = listeners.split_off(&(sequence + 1));
```

Comment:
Good find. Too bad there's no `split_off_reverse` or this could be quite elegant! I think you've found the best option, I'd only suggest renaming `b` something like `outstanding_listeners`.